pkg/compactor/block_upload.go

```diff
@@ -262,29 +285,73 @@ func (c *MultitenantCompactor) UploadBlockFile(w http.ResponseWriter, r *http.Re
 	w.WriteHeader(http.StatusOK)
 }
 
-func (c *MultitenantCompactor) completeBlockUpload(ctx context.Context, logger log.Logger, userBkt objstore.Bucket, blockID ulid.ULID, meta metadata.Meta) error {
+func (c *MultitenantCompactor) validateAndCompleteBlockUpload(logger log.Logger, userBkt objstore.Bucket, blockID ulid.ULID, meta *metadata.Meta) {
 	level.Debug(logger).Log("msg", "completing block upload", "files", len(meta.Thanos.Files))
 
-	// Upload meta file so block is considered complete
+	ch := make(chan struct{})
+	var wg sync.WaitGroup
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// start a go routine that updates the validation file's timestamp every heartbeat interval
+	wg.Add(1)
+	go c.periodicValidationUpdater(ctx, logger, blockID, userBkt, ch, &wg, cancel)
+
+	if err := c.validateBlock(ctx, blockID, userBkt, meta); err != nil {
+		level.Error(logger).Log("msg", "error while validating block", "err", err)
+		close(ch)
+		wg.Wait()
+		if !errors.Is(err, context.Canceled) {
+			err := c.uploadValidationWithError(ctx, blockID, userBkt, err.Error())
+			if err != nil {
+				level.Error(logger).Log("msg", "error updating validation file after failed block validation", "err", err)
+			}
+		}
+		return
+	}
+
+	close(ch)
+	// use waitgroup to ensure validation ts update is complete
+	wg.Wait()
+
+	if err := c.markBlockComplete(ctx, logger, userBkt, blockID, meta); err != nil {
+		if !errors.Is(err, context.Canceled) {
```

Comment:
If we got here, we should't get `context.Canceled`, because only `periodicValidationUpdater` can cancel the context, but we stopped that goroutine already. (And if it *did* cancel the context, we would most likely get error from `validateBlock` instead)

I would just remove this check to simplify the code.

Reply:
Removed this context check as well.